src/main/java/com/yahoo/bullet/storm/batching/BatchManager.java

```diff
@@ -0,0 +1,249 @@
+ * resized. It is not a guarantee on the max size of a batch.
+ *
+ * The batch manager supports GZIP compression. Batches are only compressed when partitions are resized or when the
+ * compressed batches are retrieved, in which case only changed batches will be re-compressed.
+ *
+ * Batches can be retrieved both partitioned and non-partitioned. These batches are separate data structures and do not
+ * reflect changes to the batch manager.
+ */
+/* Exposed for testing only. */
+@Getter(AccessLevel.PACKAGE)
+@Slf4j
+public class BatchManager<T> {
+    private final List<Partition<T>> partitions;
+    private final int partitionCount;
+    private final boolean batchCompressEnable;
+
+    /**
+     * Creates a BatchManager with the given batch size and partition count.
+     *
+     * @param batchSize The approximate maximum number of elements in a batch.
+     * @param partitionCount The number of partitions to use.
+     * @param batchCompressEnable Whether or not to serialize and compress batches.
+     */
+    public BatchManager(int batchSize, int partitionCount, boolean batchCompressEnable) {
+        log.info("Creating BatchManager with batch size {}, partition count {}, and batch compression enabled {}.",
+                batchSize, partitionCount, batchCompressEnable);
+
+        if (partitionCount < 1) {
+            throw new RuntimeException("Partition count must be greater than 0. The parameter given was " + partitionCount);
+        }
+
+        partitions = new ArrayList<>(partitionCount);
+        for (int i = 0; i < partitionCount; i++) {
+            partitions.add(new Partition<>(i, batchSize, batchCompressEnable));
+        }
+
+        this.partitionCount = partitionCount;
+        this.batchCompressEnable = batchCompressEnable;
+    }
+
+    /**
+     * Get all batches as a list of batches.
+     *
+     * @return A list containing all batches.
+     */
+    public List<Map<String, T>> getBatches() {
+        return partitions.stream().map(Partition::getImmutableBatches).flatMap(Collection::stream).collect(Collectors.toList());
+    }
+
+    /**
+     * Get all batches from the given partition.
+     *
+     * @param index The partition index.
+     * @return All the batches for the given partition as a {@link List} of {@link Map}.
+     */
+    public List<Map<String, T>> getBatchesForPartition(int index) {
+        return partitions.get(index).getImmutableBatches();
+    }
+
+    /**
+     * Get all batches as lists of batches mapped by partition.
+     *
+     * @return A map containing all batches mapped by partition.
+     */
+    public Map<Integer, List<Map<String, T>>> getPartitionedBatches() {
+        return partitions.stream().collect(Collectors.toMap(Partition::getId, Partition::getImmutableBatches));
+    }
+
+    /**
+     * Get all batches back as a list of compressed bytes.
+     *
+     * @return A list containing all batches compressed.
+     */
+    public List<byte[]> getCompressedBatches() {
```

Comment:
Do these need to be public? It seems only Partition uses it? I think only getBatches and getBatchesForPartition are actually used.

Reply:
Actually I think these methods don't need to exist since Partition handles doing the compression.

Reply:
They're not used atm since compressed batches isn't supported in replay yet... but I will add that in this pr.